tests/d3d12_mesh_shader.c

```diff
@@ -0,0 +1,2326 @@
+    hr = create_root_signature(device, &root_signature_desc, &root_signature);
+    ok(SUCCEEDED(hr), "Failed to create root signature, hr %#x.\n", hr);
+
+    /* On the most basic level, only a root signature and mesh shader are required */
+    ms_only_pipeline_desc.root_signature = root_signature_subobject;
+    ms_only_pipeline_desc.root_signature.root_signature = root_signature;
+    ms_only_pipeline_desc.ms = ms_subobject;
+
+    hr = create_pipeline_state_from_stream(device2, &ms_only_pipeline_desc, &pipeline_state);
+    ok(SUCCEEDED(hr), "Failed to create pipeline, hr %#x.\n", hr);
+    ID3D12PipelineState_Release(pipeline_state);
+
+    /* Mixing mesh shaders and any part of the old geometry pipelines is not allowed */
+    ms_vs_pipeline_desc.root_signature = root_signature_subobject;
+    ms_vs_pipeline_desc.root_signature.root_signature = root_signature;
+    ms_vs_pipeline_desc.ms = ms_subobject;
+    ms_vs_pipeline_desc.vs = vs_subobject;
+    ms_vs_pipeline_desc.primitive_topology = primitive_topology_subobject;
+    ms_vs_pipeline_desc.input_layout = input_layout_subobject;
+    ms_vs_pipeline_desc.rasterizer = rasterizer_subobject;
+
+    hr = create_pipeline_state_from_stream(device2, &ms_vs_pipeline_desc, &pipeline_state);
+    ok(hr == E_INVALIDARG, "Unexpected result for pipeline creation, hr %#x.\n", hr);
+
+    /* Pixel shaders require a non-trivial mesh shader */
+    ms_ps_pipeline_desc.root_signature = root_signature_subobject;
+    ms_ps_pipeline_desc.root_signature.root_signature = root_signature;
+    ms_ps_pipeline_desc.ms = ms_subobject;
+    ms_ps_pipeline_desc.ps = ps_subobject;
+    ms_ps_pipeline_desc.sample_desc = sample_desc_subobject;
+    ms_ps_pipeline_desc.rasterizer = rasterizer_subobject;
+
+    hr = create_pipeline_state_from_stream(device2, &ms_ps_pipeline_desc, &pipeline_state);
+    ok(hr == E_INVALIDARG, "Unexpected result for pipeline creation, hr %#x.\n", hr);
+
+    ms_ps_pipeline_desc.ms = ms_nontrivial_subobject;
+
+    hr = create_pipeline_state_from_stream(device2, &ms_ps_pipeline_desc, &pipeline_state);
+    ok(SUCCEEDED(hr), "Failed to create pipeline, hr %#x.\n", hr);
+    ID3D12PipelineState_Release(pipeline_state);
+
+    /* Input assembly and primitive topology are ignored */
+    ms_ia_pipeline_desc.root_signature = root_signature_subobject;
+    ms_ia_pipeline_desc.root_signature.root_signature = root_signature;
+    ms_ia_pipeline_desc.ms = ms_subobject;
+    ms_ia_pipeline_desc.primitive_topology = primitive_topology_subobject;
+    ms_ia_pipeline_desc.input_layout = input_layout_subobject;
+
+    hr = create_pipeline_state_from_stream(device2, &ms_ia_pipeline_desc, &pipeline_state);
+    ok(SUCCEEDED(hr), "Failed to create pipeline, hr %#x.\n", hr);
+    ID3D12PipelineState_Release(pipeline_state);
+
+    ID3D12RootSignature_Release(root_signature);
+    ID3D12Device2_Release(device2);
+    ID3D12Device_Release(device);
```

Comment:
Missing test context destroy?

Reply:
There is no context here, we just create a raw device for this test.

Reply:
True.